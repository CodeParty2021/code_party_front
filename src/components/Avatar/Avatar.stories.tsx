import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Avatar } from "./Avatar";

export default {
  title: "components/Avatar",
  component: Avatar,
} as ComponentMeta<typeof Avatar>;

const Template: ComponentStory<typeof Avatar> = (args) => <Avatar {...args} />;

export const Default = Template.bind({});
Default.args = {
  userPhotoUrl: undefined,
  color: "default",
  type: "user",
};

export const Turquoise = Template.bind({});
Turquoise.args = {
  ...Default.args,
  color: "turquoise",
};

export const Leaf = Template.bind({});
Leaf.args = {
  ...Default.args,
  color: "leaf",
};

export const Orange = Template.bind({});
Orange.args = {
  ...Default.args,
  color: "orange",
};

export const Pink = Template.bind({});
Pink.args = {
  ...Default.args,
  color: "pink",
};

export const HasIcon = Template.bind({});
HasIcon.args = {
  ...Default.args,
  userPhotoUrl: "/logo192.png",
};

export const User = Template.bind({});
User.args = {
  ...Default.args,
  type: "user",
};

export const Anonymous = Template.bind({});
Anonymous.args = {
  ...Default.args,
  type: "anonymous",
};

export const Robot = Template.bind({});
Robot.args = {
  ...Default.args,
  type: "robot",
};
