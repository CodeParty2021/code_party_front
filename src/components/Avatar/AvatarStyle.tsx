import styled, { css } from "styled-components";
import { ReactSVG } from "react-svg";
import { avatar } from "styles/colors";

export type AvatarStyleProps = {
  color?: "default" | "turquoise" | "leaf" | "orange" | "pink" | "gray";
  type?: "user" | "anonymous" | "robot";
};

type PrivateProps = {
  imageUrl?: string;
};

export const AvatarDummyStyle = styled(ReactSVG)`
  display: block;
  width: 32px;
  height: 38px;
`;

const defaultStyle = css`
  display: block;
  width: 54px;
  height: 54px;
  /* ボーダーからはみ出ないように中身は1px内側に描画する */
  padding: 1px;

  position: relative;

  .avatar_main {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    padding: 2px;
    overflow: hidden;
    width: 100%;
    height: 100%;
    border-radius: 50%;
    background-color: ${avatar.default.background};

    ${AvatarDummyStyle} {
      align-self: flex-end;
      fill: ${avatar.default.avatar};
    }

    .avatar_img {
      object-fit: cover;
      width: 100%;
      height: 100%;
    }
  }

  /* ボーダーは上から重ねる */
  .avatar_circle {
    position: absolute;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    border-radius: 50%;
    border: 3px solid ${avatar.default.border};
  }
`;

const turquoiseStyle = css`
  .avatar_main {
    background-color: ${avatar.turquoise.background};

    ${AvatarDummyStyle} {
      fill: ${avatar.turquoise.avatar};
    }
  }

  .avatar_circle {
    border-color: ${avatar.turquoise.border};
  }
`;

const leafStyle = css`
  .avatar_main {
    background-color: ${avatar.leaf.background};

    ${AvatarDummyStyle} {
      fill: ${avatar.leaf.avatar};
    }
  }

  .avatar_circle {
    border-color: ${avatar.leaf.border};
  }
`;

const orangeStyle = css`
  .avatar_main {
    background-color: ${avatar.orange.background};

    ${AvatarDummyStyle} {
      fill: ${avatar.orange.avatar};
    }
  }

  .avatar_circle {
    border-color: ${avatar.orange.border};
  }
`;

const pinkStyle = css`
  .avatar_main {
    background-color: ${avatar.pink.background};

    ${AvatarDummyStyle} {
      fill: ${avatar.pink.avatar};
    }
  }

  .avatar_circle {
    border-color: ${avatar.pink.border};
  }
`;

const grayStyle = css`
  border-color: ${avatar.gray.border};
  background-color: ${avatar.gray.background};
`;

const userStyle = css``;

const anonymousStyle = css`
  .avatar_main {
    background-color: ${avatar.anonymous.background};

    ${AvatarDummyStyle} {
      fill: ${avatar.anonymous.avatar};
    }
  }

  .avatar_circle {
    border-color: ${avatar.anonymous.border};
  }
`;

const imageStyle = css<PrivateProps>``;

const noImageStyle = css`
  .avatar_main {
    background: ${avatar.default.noImage.background};
  }
`;

const robotStyle = css`
  .avatar_main {
    background-color: ${avatar.robot.background};
  }

  .avatar_circle {
    border-color: ${avatar.robot.border};
  }
`;

export const AvatarStyle = styled.div<AvatarStyleProps & PrivateProps>`
  ${defaultStyle}

  ${({ color }) => color === "turquoise" && turquoiseStyle}
  ${({ color }) => color === "leaf" && leafStyle}
  ${({ color }) => color === "orange" && orangeStyle}
  ${({ color }) => color === "pink" && pinkStyle}

  ${({ imageUrl }) => imageUrl && imageStyle}
  ${({ color, imageUrl }) => color === "default" && !imageUrl && noImageStyle}

  ${({ type }) => type === "user" && userStyle}
  ${({ type }) => type === "anonymous" && anonymousStyle}
  ${({ type }) => type === "robot" && robotStyle}

  ${({ color }) => color === "gray" && grayStyle}
`;

AvatarStyle.defaultProps = {
  color: "default",
  type: "anonymous",
};
