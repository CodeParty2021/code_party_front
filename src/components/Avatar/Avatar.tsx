import React from "react";
import { AvatarStyle, AvatarStyleProps, AvatarDummyStyle } from "./AvatarStyle";
import { AlgoHead } from "components/Character/Algo/Head/AlgoHead";

type Props = AvatarStyleProps & {
  userPhotoUrl?: string;
};

export const Avatar: React.FC<Props> = ({ userPhotoUrl, ...styleProps }) => {
  return (
    <AvatarStyle imageUrl={userPhotoUrl} {...styleProps}>
      {userPhotoUrl ? undefined : styleProps.type === "robot" ? (
        <AlgoHead color="blue" width="34px" height="28px" />
      ) : (
        !userPhotoUrl && (
          <AvatarDummyStyle src="/img/avatar_dummy.svg" wrapper="svg" />
        )
      )}
    </AvatarStyle>
  );
};
