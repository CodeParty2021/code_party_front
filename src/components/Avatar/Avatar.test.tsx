import React from "react";
import { Avatar } from "./Avatar";
import "jest-styled-components";
import { render } from "enzyme";

describe("<Avatar />", () => {
  it("snapshot test", () => {
    const wrapper = render(<Avatar />);

    expect(wrapper).toMatchSnapshot();
  });

  it("render test user turquoise", () => {
    expect(
      render(<Avatar type="user" color="turquoise" />)
    ).toBeTruthy();
  });

  it("render test user leaf", () => {
    expect(
      render(<Avatar type="user" color="leaf" />)
    ).toBeTruthy();
  });

  it("render test user orange", () => {
    expect(
      render(<Avatar type="user" color="orange" />)
    ).toBeTruthy();
  });

  it("render test user pink", () => {
    expect(
      render(<Avatar type="user" color="pink" />)
    ).toBeTruthy();
  });

  it("render test user hasImage", () => {
    expect(
      render(<Avatar type="user" userPhotoUrl="/test.png" />)
    ).toBeTruthy();
  });

  it("render test anonymous", () => {
    expect(
      render(<Avatar type="anonymous" />)
    ).toBeTruthy();
  });

  it("render test robot", () => {
    expect(
      render(<Avatar type="robot" />)
    ).toBeTruthy();
  });
});
