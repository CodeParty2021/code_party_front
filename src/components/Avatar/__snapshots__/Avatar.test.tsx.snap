// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> snapshot test 1`] = `
.c2 {
  width: 32px;
  height: 38px;
}

.c0 {
  display: block;
  width: 54px;
  height: 54px;
  border-radius: 50%;
  border: 3px solid #FFFFFF;
  background-color: #EAEBF1;
  background-size: cover;
  overflow: hidden;
  position: relative;
  border-color: #FFFFFF;
  background-color: #EAEBF1;
}

.c0 .c1 {
  position: absolute;
  bottom: 0px;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c0 .c1 {
  fill: #8A9CFF;
}

.c0 .c1 {
  fill: #A2AAC2;
}

<div
  class="c0"
  color="default"
  type="anonymous"
>
  <svg
    class="c1 c2"
    xlink="http://www.w3.org/1999/xlink"
    xmlns="http://www.w3.org/2000/svg"
  />
</div>
`;
